pkg/controllers/disruption/controller.go

```diff
@@ -194,10 +203,42 @@ func (c *Controller) executeCommand(ctx context.Context, m Method, cmd Command,
 		return multierr.Append(fmt.Errorf("tainting nodes (command-id: %s), %w", commandID, err), state.RequireNoScheduleTaint(ctx, c.kubeClient, false, stateNodes...))
 	}
 
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("interrupted (command-id: %s)", commandID)
+	case <-c.clock.After(commandValidationDelay):
+	}
+
+	// Revalidate the proposed command after the node has been tainted.
+	// This addresses a race conditions where pods can schedule to the candidate nodes after the command has been finalized, but before the candidates were tainted.
+	candidates, err := GetCandidates(ctx, c.cluster, c.kubeClient, c.recorder, c.clock, c.cloudProvider, m.ShouldDisrupt, c.queue)
+	if err != nil {
+		return fmt.Errorf("validating command, determining candidates, %w", err)
+	}
+	disruptionBudgetMapping, err := BuildDisruptionBudgets(ctx, c.cluster, c.clock, c.kubeClient, c.recorder)
+	if err != nil {
+		return fmt.Errorf("validating command, building disruption budgets, %w", err)
+	}
+	validatedCommand := m.ValidateCommand(ctx, cmd, disruptionBudgetMapping, candidates...)
+	if len(validatedCommand.candidates) == 0 {
+		logging.FromContext(ctx).Infof("abandoning disruption attempt, 0 out of %d nodes are still valid candidates", len(cmd.candidates))
+		return state.RequireNoScheduleTaint(ctx, c.kubeClient, false, stateNodes...)
+	}
+	if len(validatedCommand.candidates) != len(cmd.candidates) {
+		invalidatedCandidates := lo.FilterMap(cmd.candidates, func(candidate *Candidate, _ int) (*state.StateNode, bool) {
+			return candidate.StateNode, !lo.ContainsBy(validatedCommand.candidates, func(c *Candidate) bool {
+				return c.Name() == candidate.Name()
+			})
+		})
+		if err = state.RequireNoScheduleTaint(ctx, c.kubeClient, false, invalidatedCandidates...); err != nil {
+			return fmt.Errorf("untainting invalidated candidates, %w", err)
+		}
+		logging.FromContext(ctx).Infof("updated disruption command after validation, disrupting via %s %s", m.Type(), validatedCommand)
+	}
```

Comment:
Should we be continuing if a subset of the candidates are no longer candidates? Let's say we had 100 candidates, and now only 1 candidate is invalid. We need to either re-do a scheduling simulation to know what the replacement should be, or just abandon the command entirely. 

For reference, we abandon the command for consolidation.

Reply:
Like you called out the decision we make here is going to differ depending on the disruption type. It was handled in the ValidateCommand function for that reason. If we want to cancel the entire command it should return a command with no candidates. Otherwise it should return a command with a reduced number of candidates and we would untaint the invalidated candidates here.